server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java

```diff
@@ -1158,117 +1258,59 @@ public static List<String> resolve(Context context, List<String> expressions) {
                 return resolvedExpressions;
             }
 
-            Set<String> result = innerResolve(context, expressions, options, metadata);
-
-            if (result == null) {
-                return expressions;
-            }
-            if (result.isEmpty() && options.allowNoIndices() == false) {
-                IndexNotFoundException infe = new IndexNotFoundException((String) null);
-                infe.setResources("index_or_alias", expressions.toArray(new String[0]));
-                throw infe;
+            ExplicitExpressionResolver.Resolved explicitsAndWildcards = ExplicitExpressionResolver.resolve(
+                context,
+                expressions,
+                options,
+                metadata
+            );
+            if (explicitsAndWildcards.wildcardsAndExclusions().isEmpty()) {
+                return new ArrayList<>(explicitsAndWildcards.explicitNames());
+            } else {
+                Set<String> expandedWildcards = innerResolve(context, explicitsAndWildcards.wildcardsAndExclusions(), options, metadata);
+                if (expandedWildcards.isEmpty() && options.allowNoIndices() == false) {
+                    IndexNotFoundException infe = new IndexNotFoundException((String) null);
+                    infe.setResources("index_or_alias", explicitsAndWildcards.wildcardsAndExclusions().toArray(new String[0]));
+                    throw infe;
+                }
+                expandedWildcards.addAll(explicitsAndWildcards.explicitNames());
+                return new ArrayList<>(expandedWildcards);
```

Comment:
Nit: As mentioned during our discussion, it might make sense to return immutable collection and change it to `Set` instead of `List`?